prod/ODWithWLS

mahditaani opened the pull request:
This is the latest version of the hyper-k od. This includes the most recent hyper-k geometry as well as fixes to the photocoverage code.

P3tru committed:
Implement a cosmic generator

Based on Super-K cosmic flux extrapolated to Hyper-K site

P3tru committed:
Add visu using QT drivers in interactive mode

P3tru committed:
Defining relevant classes

Embryo classes and methods for combined PMTs collection stacking action

P3tru committed:
Concatenate PMTs QE info in one vector

Working. Need to add that in stacking action now

P3tru committed:
Add improved application of QE in StackingAction

Add a new method to apply QE inside StackingAction for different PMT collection

P3tru committed:
Fix stupid mistake cause I can't read compiler warnings

P3tru committed:
Merge branch 'feature/AddCaveWithTyvek' into feature/WLSPlates

P3tru committed:
Add WLS classes

P3tru committed:
Add WLS material and set optical properties

P3tru committed:
Implement WLS Geometry

P3tru committed:
Add recovery of PMT information

P3tru committed:
Add WLS absorption info inside stacking Action

The stacking action takes into account the QE of the PMTs and the absorption of the WLS plates

P3tru committed:
Merge branch 'feature/ODDigi' into prod/OD

P3tru committed:
Correct stupid bug

P3tru committed:
Merge branch 'feature/QTVis' into prod/OD

P3tru committed:
Add OD messenger and small PMTs option

P3tru committed:
Debug photons info

P3tru committed:
Add Qt macro for debugging purposes

P3tru committed:
Added Mahdi WLS plates response function

P3tru committed:
Very stupid mistake from Stephane

P3tru committed:
Fix overlapping bug

P3tru committed:
Clean WLS materials name

P3tru committed:
HyperKOD prod

mahditaani committed:
Create the emission spectrum for the WLS plates outside of the SteppingAction class. Will be created once and used in SteppingAction to increase speed of simulation.

P3tru committed:
Merge pull request #4 from mahditaani/hotfix/WLSPlates

Create the emission spectrum for the WLS plates outside of the Steppi…

P3tru committed:
Externalize parameters from stepping action

And more tests

P3tru committed:
Merge remote-tracking branch 'origin/hotfix/WLSPlates' into hotfix/WLSPlates

P3tru committed:
Try new geom for OD

P3tru committed:
Correct inline colour

P3tru committed:
Merge branch 'develop' into hotfix/WLSPlates

P3tru committed:
Change random gen for compatibility with c++98

P3tru committed:
Now OK with old gcc and faster TRandom

P3tru committed:
Might work now ? Seeing 40% of shifted photon inside the plates

P3tru committed:
Merge branch 'feature/CosmicsGen' into hotfix/WLSPlates

P3tru committed:
Add option to deactivate WLS plates in messenger

P3tru committed:
Clean stepping action and add collection of photons in WLSplates

P3tru committed:
Edit macro for OD analysis

P3tru committed:
Revert federico commit

P3tru committed:
Correct visuPhotons macro

P3tru committed:
Add more comments

P3tru committed:
Fix Lilia's bug I hope !

mahditaani committed:
Merge remote-tracking branch 'stephane/hotfix/WLSPlates' into hotfix/WLSPlates

mahditaani committed:
Added a script to read the OD produce a root file. Also added a script to load the libraries.

P3tru committed:
Try an other other fix... And edit OD.mac and read_OD

P3tru committed:
Merge pull request #5 from mahditaani/hotfix/WLSPlates

Hotfix/wls plates

P3tru committed:
Deactivate ID PMT to speed up cosmic simu

P3tru committed:
AAAAaaaaAAAAgh need to look at my macro file before commiting

P3tru committed:
Add simple macro to compute OD trigger efficiency

P3tru committed:
Fix nPMTs bug for small option

P3tru committed:
Cleaning the branch

P3tru committed:
Fix bug with PMTs size over WLS container

P3tru committed:
clean branch

P3tru committed:
Clean macro file and analyse scripts

mahditaani committed:
Vector files can now be used. Changed variable names to avoid conflicts.

P3tru committed:
Merge pull request #6 from mahditaani/prod/ODWithWLS

Vector files can now be used. Changed variable names to avoid conflicts.

mahditaani committed:
Added Some OD PMT Info so the positions of the PMT can be accessed etc.

mahditaani committed:
Fixed minor error in comments.

P3tru committed:
Add random seed for gene

P3tru committed:
Fix few bugs for photons collected inside photons tree

mahditaani committed:
Fixed bug in WLS code. Code requested a creatorprocess for optical photons; creator processes don't exist for initial particles.

P3tru committed:
Merge pull request #8 from mahditaani/prod/ODWithWLS

Fixed bug in WLS code. Code requested a creatorprocess for optical ph…

P3tru committed:
Merge pull request #7 from mahditaani/AddODInfo

Add od info

P3tru committed:
Merge branch 'develop' into prod/ODWithWLS

P3tru committed:
Merge remote-tracking branch 'origin/prod/ODWithWLS' into prod/ODWithWLS

mahditaani committed:
Fixed the Tyvek reflectivity. Changed the optical border surfaces (reflectivity defined between the border of two physical volumes) to skin surfaces (reflectivity defined as the skin of a logical volume). Also changed the mother volumes for the Tyvek so that it is always in the water volume. Changed some of the logical Tyvek volumes to the correct shape.

P3tru committed:
Fix container issue

P3tru committed:
Merge pull request #9 from mahditaani/fixTyvek

Fixed the Tyvek reflectivity. Changed the optical border surfaces (re…

P3tru committed:
Merge remote-tracking branch 'origin/prod/ODWithWLS' into prod/ODWithWLS

P3tru committed:
Fix overlaps

mahditaani committed:
Changed Tyvek material properties to include a different reflection model.

P3tru committed:
Fix PMT placement bug

P3tru committed:
Leave materials borders for container/structure interface in the OD

mahditaani committed:
Changed the reflectivity table for the Tyvek. New table based on plot found in https://www.osti.gov/servlets/purl/1184400 (US Department Of Energy OSTI).

P3tru committed:
Merge pull request #10 from mahditaani/prod/ODWithWLS

Changed Tyvek material properties to include a different reflection m…

P3tru committed:
Merge remote-tracking branch 'origin/prod/ODWithWLS' into prod/ODWithWLS

P3tru committed:
Remove logical border surfaces for OD interfaces

P3tru committed:
Update WLS plate smearing time

P3tru committed:
Sanitize stepping action

P3tru committed:
Sanitize steppingaction

P3tru committed:
Sanitize Tyvek and WLS optical props

P3tru committed:
Change float to double inside WLS props

P3tru committed:
Extend optical properties for WLS

P3tru committed:
Collect photons entering plate without killing them

P3tru committed:
Add fit in read_OD script

P3tru committed:
Update HKComputing and adding tyvek jobs

P3tru committed:
Push latest read_OD modifs

P3tru committed:
Addition to HKComputing

P3tru committed:
Fix bin types in read_OD.C

P3tru committed:
Create cosmics jobs macro

P3tru committed:
Update ABS spectrum Eljen EJ-286 plate

P3tru committed:
Add edits for WLS physics list and some bug fixes

P3tru committed:
Add WLS cladding. Fix a few minor bugs

P3tru committed:
Fix bug to save cosmic generator information

P3tru committed:
Merge remote-tracking branch 'origin/prod/ODWithWLS' into prod/ODWithWLS

mahditaani committed:
Fixed a bug in the OD photocoverage. Now the photocoverage parameter sets the correct number of PMTs in the OD.

mahditaani committed:
Updated the HK geometry to the most recent cavern values.

mahditaani committed:
Merge branch 'prod/ODWithWLS' of https://github.com/P3tru/WCSim into prod/ODWithWLS

mahditaani committed:
Corrected silly mistake. Now up to date HK geometry.

mahditaani committed:
Fixed minor bug with PMT coverage.

mahditaani committed:
Corrected PMT placement when extra tower is needed.

mahditaani committed:
Corrected the spacing of PMTs and the width of WLS containers.

mahditaani committed:
Added a variable to offset PMTs in the border ring. This stops them from overlapping with the slanted angle. The offset also reduces the overall height of the cell so PMTs are slightly closer together so that they can all fit.

mahditaani committed:
Fixed overlap in OD geometry. OD PMTs were being placed in a slightly off angle and slightly displaced in one direction.

federiconova committed:
add cladding; add cladding reflectivity parameter; inactivate ID; correct eljen density; hard-coded tyvek reflectivity; build WLS plate

federiconova committed:
missing file

federiconova committed:
OD.mac

federiconova committed:
pass tyvek reflectivity and cladding reflectivity as parameters

federiconova committed:
better flow

federiconova committed:
option to add plate to cathode; place plate correctly; comment unused materials; set plate parameters (abslength, updated wlsabslength, component); CreateCombinedPMTQE in updateodgeo; rm or extend hardcoded limits on photons propagation; set arbitrary ratio to 1

federiconova committed:
layer of water under WLS plate

federiconova committed:
remove option to add plate to photocathode as union volume

federiconova committed:
default build WLS cladding

federiconova committed:
restore ratio in PMTs as recommended by software group

mahditaani committed:
Implemented photon time fix from pull request: https://github.com/WCSim/WCSim/pull/259

mahditaani committed:
Changed the number of horizontal PMTs to match the real design of HyperK. Previous cell size was so small that increasing the number of OD pmts by 1 per cell greatly increases the number of total PMTs. A lerger cell makes fewer total cells giving us a greater control over the number of total PMTs.

mahditaani committed:
Barrel Cells must have fixed number of PMTs, this gives them a slightly different photocoverage than the end caps. After the Barrel PMTs are placed, a new photocoverage is calculated for the endcaps. This gives the detector a uniform coverage close to the suggested photocoverage value.

federiconova committed:
add 10k OD geometry

federiconova committed:
fix mac files

federiconova committed:
fix position of PMTs with WLS plate to resole photons getting lost

mahditaani committed:
Merge branch 'feature/add-WLS-plates' of https://github.com/federiconova/WCSim into prod/ODWithWLS

mahditaani committed:
Merge branch 'TestingGeom' of github.com:mahditaani/WCSim into prod/ODWithWLS

mahditaani committed:
Made changes to allow for finer control of the OD coverage. Updated the 10k PMT macros to give a uniform coverage throughout the OD.

mahditaani committed:
Updated HK geometry dimensions which now account for the concrete and steel outer layers.

mahditaani committed:
Updated sample OD read script to be able to correctly read the OD geometry (as well as the ID)

tdealtry commented:
Thanks for this @mahditaani (& co) We're planning to start reviewing this PR soon, so can you let us know if it's ready to go?
(Note that there are conflicts that need fixing anyway)

tdealtry committed:
Merge branch 'develop' of github.com:WCSim/WCSim into prod/ODWithWLS

Conflicts:
	.travis.yml
	Dockerfile
	Dockerfile.base
	README.md
	WCSim.cc
	WCSim.mac
	include/WCSimDetectorConstruction.hh
	include/WCSimPMTObject.hh
	include/WCSimPrimaryGeneratorAction.hh
	include/WCSimPrimaryGeneratorMessenger.hh
	include/WCSimRootEvent.hh
	include/WCSimRootGeom.hh
	include/WCSimWCDigi.hh
	include/jhfNtuple.h
	sample-root-scripts/read_OD.C
	src/WCSimConstructCylinder.cc
	src/WCSimConstructGeometryTables.cc
	src/WCSimConstructMaterials.cc
	src/WCSimConstructPMT.cc
	src/WCSimDetectorMessenger.cc
	src/WCSimEventAction.cc
	src/WCSimPMTObject.cc
	src/WCSimPhysicsListFactory.cc
	src/WCSimPrimaryGeneratorAction.cc
	src/WCSimPrimaryGeneratorMessenger.cc
	src/WCSimRunAction.cc
	src/WCSimStackingAction.cc
	src/WCSimTuningMessenger.cc

tdealtry commented:
@mahditaani I've done the merge, but can't push it to your branch. So can you pull the [merge commit](https://github.com/tdealtry/WCSim/commit/ea929653e038596a3c1a5d096316b305efd18d21) from my branch? Something like
```bash
#make sure you're on the right branch
git checkout prod/ODWithWLS
#grab my changes
git remote add tom git@github.com:tdealtry/WCSim.git
git pull tom prod/ODWithWLS
#push them to your repo
git push origin prod/ODWithWLS
```

mahditaani commented:
Hi Tom,

Thank you!

I will do it and run some checks.

All the best,
Mahdi

From: Tom Dealtry ***@***.***>
Sent: 14 May 2021 12:21
To: WCSim/WCSim ***@***.***>
Cc: Taani, Mahdi ***@***.***>; Mention ***@***.***>
Subject: Re: [WCSim/WCSim] prod/ODWithWLS (#297)


@mahditaani<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fmahditaani&data=04%7C01%7Cmahdi.1.taani%40kcl.ac.uk%7Cd47e5684b636476d458608d916ca5f87%7C8370cf1416f34c16b83c724071654356%7C0%7C0%7C637565880703806631%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=5AW72OKo%2FzvIDw22W2tmLMdAbg05f0cAOLMzgLOORvs%3D&reserved=0> I've done the merge, but can't push it to your branch. So can you pull the merge commit<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Ftdealtry%2FWCSim%2Fcommit%2Fea929653e038596a3c1a5d096316b305efd18d21&data=04%7C01%7Cmahdi.1.taani%40kcl.ac.uk%7Cd47e5684b636476d458608d916ca5f87%7C8370cf1416f34c16b83c724071654356%7C0%7C0%7C637565880703816590%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=nv2Qx%2BHNTgzRWXZ10wq03HZW0DMH%2B9N8Z7HuESY%2FiOQ%3D&reserved=0> from my branch? Something like

#make sure you're on the right branch

git checkout prod/ODWithWLS

#grab my changes

git remote add tom ***@***.******@***.***:tdealtry/WCSim.git>

git pull tom prod/ODWithWLS

#push them to your repo

git push origin prod/ODWithWLS

-
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2FWCSim%2FWCSim%2Fpull%2F297%23issuecomment-841182711&data=04%7C01%7Cmahdi.1.taani%40kcl.ac.uk%7Cd47e5684b636476d458608d916ca5f87%7C8370cf1416f34c16b83c724071654356%7C0%7C0%7C637565880703816590%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=6wdmDQk8p%2FlZXHjsdQeyxfJ9VaTNYaPy8Zl6RRBHApE%3D&reserved=0>, or unsubscribe<https://eur03.safelinks.protection.outlook.com/?url=https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FAFJUJBRHN46B3W3LLA7I5B3TNUBSJANCNFSM4SFWPQTA&data=04%7C01%7Cmahdi.1.taani%40kcl.ac.uk%7Cd47e5684b636476d458608d916ca5f87%7C8370cf1416f34c16b83c724071654356%7C0%7C0%7C637565880703826541%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C1000&sdata=JxC69t9i8676QyKqRLDqeIG6PC%2FlrP7pFKC8KlpJXcU%3D&reserved=0>.

mahditaani commented:
@tdealtry I tried to run this and while the code compiles it does not build an OD geometry (aborts when you try to run OD_10k.mac). I am looking into this to see what the issue is.

mahditaani committed:
Revert HK geometry (without OD) to old dimensions. 

New HK dimensions create many stuck track warnings so reverted from the new geometry back to the old geometry.

mahditaani committed:
Merge branch 'prod/ODWithWLS' of github.com:tdealtry/WCSim into prod/ODWithWLS

mahditaani committed:
Fixed bug preventing OD from being created.

mahditaani committed:
Turned on ID PMTs by default.

The disabled ID PMTs cause automated CI tests to fail. Set to enabled by default.

tdealtry commented on HKComputing/Create1TeVUpMuJobs.py:
```diff
@@ -0,0 +1,53 @@
+from HKODWCSimMacro import HKODWCSimMacro
```
Could you remove the files in this folder from this PR? And then submit a new PR with them?
This is a very much decoupled feature, and would be good to get more peoples' input on. This kind of mac file building & job submission is very useful, but since there isn't an official script in WCSim already, has been invented many times by many people. We should converge on one (at first glance, this looks nice)

tdealtry commented on HyperKODComputing/CreateJobs.py:
```diff
@@ -0,0 +1,22 @@
+import sys
```
similarly please put this folders' files in the new PR

tdealtry commented on MuonFlux-HyperK-ThetaPhi.dat:
```diff
@@ -0,0 +1,18000 @@
+0 0 0.005 0 0.01 1 0 2 1.45271e-12 452.074
```
This file should be documented - both it's format, and where the data has come from (a paper? some other software?)
Maybe comment lines e.g. starting with a `#`? Maybe in a README? Maybe in `WCSim.mac`?

I'd also be in favour of moving this file to a new folder `$WCSIMDIR/data` (or `inputs/` @eosullivan what do you think?), to keep $WCSIMDIR clean. We can then store in there other useful data files e.g. biasprofile.dat

mahditaani replied:
I really like this suggestion. If there are other inputs that other groups are creating it would be useful to have a data folder. I will put the file in there for now.

tdealtry commented on WCSim.mac:
```diff
@@ -172,6 +172,12 @@
+# Muons are generated according to this flux and their position is extrapolated outside the tank
+# Useful to study muons tracks in both OD and ID
+#/mygen/generator cosmics
+
```
It's good practice to show the existence of all options in `WCSim.mac`
```suggestion
#/mygen/cosmicsfile data/MuonFlux-HyperK-ThetaPhi.dat 

```
although it looks like the default in the Messenger is inputcosmicfile (which won't exist) so perhaps it should be uncommented

mahditaani replied:
I have changed the code a little so it correctly points to the default file data/MuonFlux-HyperK-ThetaPhi.dat  and so I have added the comment to WCSim.mac as you have suggested.

tdealtry commented on src/WCSimPrimaryGeneratorMessenger.cc:
```diff
@@ -13,20 +13,26 @@ WCSimPrimaryGeneratorMessenger::WCSimPrimaryGeneratorMessenger(WCSimPrimaryGener
+  fileNameCmdCosmics->SetGuidance("Select the file of cosmics.");
+  fileNameCmdCosmics->SetGuidance(" Enter the file name of the cosmics file");
+  fileNameCmdCosmics->SetParameterName("fileName",true);
+  fileNameCmdCosmics->SetDefaultValue("inputcosmicfile");
```
```suggestion
  fileNameCmdCosmics->SetDefaultValue("data/MuonFlux-HyperK-ThetaPhi.dat");
```

mahditaani replied:
Changed the default value so that it is correct. As you have suggested.

tdealtry commented on src/WCSimWCSD.cc:
```diff
@@ -160,24 +160,24 @@ G4bool WCSimWCSD::ProcessHits(G4Step* aStep, G4TouchableHistory*)
-    maxQE = fdet->GetPMTQE(WCCollectionName,wavelength,0,240,660,ratio);
-    photonQE = fdet->GetPMTQE(volumeName, wavelength,1,240,660,ratio);
+    // maxQE = fdet->GetPMTQE(WCIDCollectionName,wavelength,0,200,700,ratio);
+    maxQE = fdet->GetPMTQE(WCCollectionName,wavelength,0,200,700,ratio);
+    photonQE = fdet->GetPMTQE(WCCollectionName, wavelength,1,200,700,ratio);
     photonQE = photonQE/maxQE;
   }else if (fdet->GetPMT_QE_Method() == 3){
     ratio = 1./(1.-0.25);
-    photonQE = fdet->GetPMTQE(volumeName, wavelength,1,240,660,ratio);
+    photonQE = fdet->GetPMTQE(WCCollectionName, wavelength,1,200,700,ratio);
```
What's the reason for changing the wavelength range here? Is it valid for all PMTs?
Also, what's the reason for using `WCCollectionName` instead of `volumeName`?

This hopefully shouldn't be affecting the travis CI (which use `Stacking_Only` or QE method 1)

mahditaani replied:
Perhaps @federiconova could comment on this?

federiconova replied:
Hi. About the argument, it seemed to me GetPMTQE always expects a collection name, not a volume name (see definition in WCSimPMTQE.cc and other calls in WCSimStackingAction.cc). Is that wrong?

About the wavelength, I guess enlarging the range is always fine? We might be interested in those extra photons.

Anyway please revert this changes if they seem problematic.

mahditaani replied:
I will change the wavelengths back as this is probably what is causing the CI tests to fail.

tdealtry commented on src/WCSimTuningParameters.cc:
```diff
@@ -19,6 +19,9 @@ WCSimTuningParameters::WCSimTuningParameters()
  tvspacing = 100.0;
  topveto = false;
 
+ WCCladdingReflectivity   = 0.90; //
```
According to the tuning parameter messenger, this is the OD WLS cladding (rather than any other cladding). Can you use a more descriptive name, such as
```suggestion
 WCODWLSCladdingReflectivity   = 0.90; //
```

mahditaani replied:
Done. See 12b24ff7c8d3984fb566ed476d87e0d474f29781

tdealtry commented on src/WCSimTrackingAction.cc:
```diff
@@ -22,11 +22,6 @@ WCSimTrackingAction::WCSimTrackingAction()
   ParticleList.insert(-321); // kaon-
   ParticleList.insert(311); // kaon0
   ParticleList.insert(-311); // kaon0 bar
```
Think this needs reverting
```suggestion
  ParticleList.insert(-311); // kaon0 bar

  ParticleList.insert(11);
  ParticleList.insert(-11);
  ParticleList.insert(12);
  ParticleList.insert(-12);
```

tdealtry commented on src/WCSimTuningMessenger.cc:
```diff
@@ -47,6 +47,16 @@ WCSimTuningMessenger::WCSimTuningMessenger(WCSimTuningParameters* WCTuningPars):
   TopVeto->SetParameterName("TopVeto",true);
   TopVeto->SetDefaultValue(0);
 
+  CommandWCCladdingReflectivity = new G4UIcmdWithADouble("/WCSim/tuning/WCCladdingReflectivity",this);
```
See my comment on WCSimTuningParameters.cc for renaming the OD WLS cladding variables to be more explicit

mahditaani replied:
Done. See 12b24ff7c8d3984fb566ed476d87e0d474f29781

tdealtry commented on include/WCSimTuningParameters.hh:
```diff
@@ -36,6 +36,12 @@ public:
 
   void SaveOptionsToOutput(WCSimRootOptions * wcopt);
 
+  G4double GetWCCladdingReflectivity() {return WCCladdingReflectivity;}
```
See my comment on WCSimTuningParameters.cc for renaming the OD WLS cladding variables/methods to be more explicit

mahditaani replied:
Done. See 12b24ff7c8d3984fb566ed476d87e0d474f29781

tdealtry commented on include/WCSimTuningMessenger.hh:
```diff
@@ -36,6 +36,8 @@ private:
   //For Top Veto - jl145
   G4UIcmdWithADouble* TVSpacing;
   G4UIcmdWithABool* TopVeto;
+  G4UIcmdWithADouble* CommandWCCladdingReflectivity;
```
See my comment on WCSimTuningParameters.cc for renaming the OD WLS cladding variables to be more explicit

mahditaani replied:
Done. See 12b24ff7c8d3984fb566ed476d87e0d474f29781

tdealtry commented on WCSim.cc:
```diff
@@ -32,6 +32,8 @@ void file_exists(const char * filename) {
 int main(int argc,char** argv)
 {
+
+  gRandom->SetSeed(0);
```
```suggestion
```
gRandom seed is now set with `/WCSim/random/seed` in the same way as the CLHEP random generator is

mahditaani replied:
Nice! Done, see 2f8abe44d7a556aa7959c6c59330988786bd25ba

tdealtry commented on include/jhfNtuple.h:
```diff
@@ -62,7 +62,7 @@ static const char* ntDesc =
 "jmu:I,"
 "jp:I,"
-"npar[0,MAX_N_PRIMARIES]:I,"
+"npar[0,50]:I,"
```
Somethings gone wrong with the merge here... can you revert all changes in this file?

mahditaani replied:
Done. See 57f135ea8e0cabdff66650875e4f1271ca03dcf9

tdealtry commented on WCSim.mac:
```diff
@@ -203,15 +209,14 @@
 #/grdm/analogueMC 0
 #/grdm/BRbias 0
 
+## change the name of the output root file, default = wcsim.root
```
I can't put this comment in the right place, but...
line 23 - can you add something like (make sure the filename is correct)
```
# if you want to use the OD, please see $WCSIMDIR/macros/OD.mac for the latest design, as well as the complete set of OD geometry options
```

mahditaani replied:
Done. See fb97cd96bc61ac50f1060fe6965a8289dd18324c

tdealtry commented on macros/OD.mac:
```diff
@@ -0,0 +1,132 @@
+# Sample setup macro with no visualization
```
Please rename this file OD_13k.mac

mahditaani replied:
I think it might be better to keep this as OD.mac because this is the default baseline design. We are currently simulating an additional 10k geometry which is why there is also an OD_10k.mac. What do you think?

tdealtry replied:
Ok that makes sense if that's the current baseline

tdealtry commented on macros/OD.mac:
```diff
@@ -0,0 +1,132 @@
+#grab the other DAQ options (thresholds, timing windows, etc.)
+/control/execute macros/daq.mac
+
+# default dark noise frequency (and conversion factor) is PMT property (NEW), set in the code.
+# Below gives possibility to overwrite nominal values, eg. to switch OFF the Dark Noise.
+#/DarkRate/SetDarkRate 0 kHz   #Turn dark noise off
+#/DarkRate/SetDarkRate 4.2 kHz #This is the value for SKI set in SKDETSIM.
+#/DarkRate/SetDarkRate 8.4 kHz #For 20 inch HPDs and Box and Line PMTs, based on High QE 20" R3600 dark rate from EGADS nov 2014
+#/DarkRate/SetDarkRate 3.0 kHz #For 12 inch HPDs and Box and Line PMTs, based on High QE 20" R3600 dark rate from EGADS nov 2014
+
+# command to multiply the dark rate.
+# Convert dark noise frequency before digitization to after digitization by setting suitable factor
+# Again, this is now a PMT property and can be overridden here
+#/DarkRate/SetConvert 1.367  #For Normal PMT
+#/DarkRate/SetConvert 1.119 #For HPDs
+#/DarkRate/SetConvert 1.126 #For Box and Line PMTs
```
Can you move the relevant parts of this to below each 
```
/DarkRate/SetDetectorElement
```
line? i.e. have to copies of the DarkRate & DarkConvert parts, in a similar way to the DarkMode parts

There's a lot of potential of confusion without this, with people (me included) attempting to turn off ID dark noise but ending up doing the OD instead (even in geometries without the OD!)

This comment applies for the other OD mac files

mahditaani replied:
There are already two copies of each command underneath the tank and od parts. I guess the comments are confusing though.

mahditaani replied:
I have modified this for clarity. See c2f0c1aa25d9c8886fb1674f386ccca2a4705f16

tdealtry commented on macros/OD__10k_OD_PMTs__no_ID__no_dark_noise__1_GeV_horizontal_muon.mac:
```diff
@@ -0,0 +1,132 @@
+# Sample setup macro with no visualization
```
Is this file required? If so, please rename it so it reflects the actual differences between this and OD_10k.mac (which seem to be a different WLS plate size, and a different spacial/energy distribution for the muons)
Having the same output filename as OD_10k.mac is potentially confusing too

mahditaani replied:
You are right, this is a redundant macro and will be removed.

tdealtry commented on macros/OD.mac:
```diff
@@ -0,0 +1,132 @@
+/WCSim/WCgeom HyperKWithOD
+
+# HyperK Outer Detector
+/WCSim/HyperKOD/PMTODRadius 3inch
+/WCSim/HyperKOD/ODLateralWaterDepth 1 m
+/WCSim/HyperKOD/ODHeightWaterDepth 2 m
+/WCSim/HyperKOD/ODDeadSpace 600 mm
+/WCSim/HyperKOD/ODTyvekSheetThickness 1 mm
+/WCSim/HyperKOD/ODWLSPlatesThickness 0.6 cm
+/WCSim/HyperKOD/ODWLSPlatesLength 30 cm
+/WCSim/HyperKOD/PMTODperCellHorizontal 0
+/WCSim/HyperKOD/PMTODperCellVertical 0
+/WCSim/HyperKOD/PMTODPercentCoverage 0.28
+/WCSim/HyperKOD/ODPMTShift 0 m
+#/WCSim/HyperKOD/DeactivateWLS
+/WCSim/HyperKOD/BuildCladding
```
Can you please comment each parameter?
This applies for the other OD mac files

mahditaani replied:
Sure, done.

tdealtry commented on macros/OD.mac:
```diff
@@ -0,0 +1,132 @@
+
+## change the name of the output root file, default = wcsim.root
+
+/WCSimIO/RootFile wcsim_13k_OD_PMTs_no_ID_no_dark_noise_horizontal_muon.root
```
This says that there's no ID. How is that achieved? How can the ID be activated? It'd be good to have an example macro with ID+OD

mahditaani replied:
This is achieved by stopping the ID PMTs from being built (using an if statement) but this is not yet changeable from the mac files so I have removed the feature and changed the filenames.

tdealtry commented on src/WCSimTuningParameters.cc:
```diff
@@ -19,6 +19,9 @@ WCSimTuningParameters::WCSimTuningParameters()
  topveto = false;
 
+ WCCladdingReflectivity   = 0.90; //
+ WCODTyvekReflectivity   = 0.97; //
```
```suggestion
 WCODTyvekReflectivity   = 0.90; //
```
the default appears to be 0.97 in WCSimTuningMessenger.cc and tuning_parameters.mac

mahditaani replied:
Good spot! yes, I will implement this suggestion.

tdealtry commented on prod.py:
```diff
@@ -0,0 +1,19 @@
+from CreateMacro import *
```
another thing for the macro/job submission PR
(FYI there appears to be something missing - grepping `CreateMacro` can't find anything - but the jobs ones exist)

tdealtry reviewed (changes requested):
Thanks for this @mahditaani 
Good spot on the part causing the OD to not build!

I've been through some of the changes (see mostly minor comments inline) but got a lot more to do tomorrow

For reference, https://github.com/WCSim/Validation/tree/master/Generate/macReference are the files run in the Travis CI, and the output is posted at https://wcsim.github.io/Validation/

tdealtry commented on include/WCSimPrimaryGeneratorAction.hh:
```diff
@@ -5,19 +5,19 @@
 #include "WCSimRootOptions.hh"
 #include "WCSimGenerator_Radioactivity.hh"
 
+#include "TH1D.h"
```
```suggestion
```
this doesn't appear to be used

mahditaani replied:
You are right, it wasn't used. Removed it. See c8b9e4448a76a9f6b26a635d1b0680429cddb160

tdealtry commented on src/WCSimPrimaryGeneratorAction.cc:
```diff
@@ -90,8 +96,66 @@ WCSimPrimaryGeneratorAction::WCSimPrimaryGeneratorAction(
   useRadioactiveEvt  	= false;
   useRadonEvt        	= false;
+
+  // Create the relevant histograms to generate muons
```
From what I can work out, the `/mygen/cosmicsfile` option doesn't work correctly
When that is called, it reads a file into `inputFile` rather than `inputFileCosmics`, and the histograms created here in the constructor aren't updated. This needs fixing

mahditaani replied:
You are right, the code was calling a single input file rather than a separate file for comics and vector data. I have fixed this see: 52e063abba1152a0a90cb1d0c7ab900b82f60f07.

tdealtry commented on src/WCSimPrimaryGeneratorAction.cc:
```diff
@@ -90,8 +96,66 @@ WCSimPrimaryGeneratorAction::WCSimPrimaryGeneratorAction(
+      hEmeanCosmics->SetBinContent(binPhi,binCos,Emean);
+    }
+
+    TFile *file = new TFile("flux.root","RECREATE");
```
```suggestion
    TFile *file = new TFile("cosmicflux.root","RECREATE");
```

mahditaani replied:
Done d48090901ca1bd3c2a2ee4f5cf0328e77755cc66

tdealtry commented on src/WCSimPrimaryGeneratorAction.cc:
```diff
@@ -90,8 +96,66 @@ WCSimPrimaryGeneratorAction::WCSimPrimaryGeneratorAction(
+    TFile *file = new TFile("flux.root","RECREATE");
+    hFluxCosmics->Write();
+    hEmeanCosmics->Write();
+    file->Close();
```
Is this required? Or was it just for debugging purposes?
Perhaps it could be an mac option if (and where) to save these histograms?

tdealtry commented on src/WCSimPrimaryGeneratorAction.cc:
```diff
@@ -288,9 +352,6 @@ void WCSimPrimaryGeneratorAction::GeneratePrimaries(G4Event* anEvent)
 
-    //To prevent occasional seg fault from an un assigned targetpdg 
-    targetpdgs[0] = 2212; //ie. proton
-
```
Please revert this

tdealtry commented on src/WCSimPrimaryGeneratorAction.cc:
```diff
@@ -332,20 +393,19 @@ void WCSimPrimaryGeneratorAction::GeneratePrimaries(G4Event* anEvent)
   }
   else if (useLaserEvt)
     {
-      targetpdgs[0] = 2212; //ie. proton 
```
please revert this

tdealtry commented on src/WCSimWLSProperties.cc:
```diff
@@ -0,0 +1,379 @@
+#define nEntries_WLS_rindex 60
+
+EljenEJ286::EljenEJ286() { SetgAbs(); SetgEm(); SethEm();}
+EljenEJ286::~EljenEJ286(){}
```
There's graphs created in the constructor (via the `SetgAbs()` etc methods). Please clean them up in the destructor

mahditaani replied:
Done, see cbd79d33abcdca16a32d19fd8ae3294921acd240

tdealtry commented on src/WCSimWLSProperties.cc:
```diff
@@ -0,0 +1,379 @@
+       0.20, 0.13, 0.081, 0.052,
+       0.03, 0.021, 0};
+
+  emissionHist = new TH1F("hist","hist", 15, 390., 530.);
```
Please check if the graph exists, and delete it if it does

mahditaani replied:
Done, see cbd79d33abcdca16a32d19fd8ae3294921acd240

tdealtry commented on src/WCSimWLSProperties.cc:
```diff
@@ -0,0 +1,379 @@
+
+void EljenEJ286::SetgEm(){
+  gEm = new TGraph();
+  const int step = 10;
```
```suggestion
```
Don't think this is used

mahditaani replied:
You're right. Good spot.

tdealtry commented on src/WCSimWLSProperties.cc:
```diff
@@ -0,0 +1,379 @@
+}
+
+void EljenEJ286::SetgEm(){
+  gEm = new TGraph();
```
It's more efficient to allocate the TGraph space required, rather than it having to do it multiple times as it's filled
```suggestion
  gEm = new TGraph(nEntries_eljen_component);
```
Also, please check if the graph exists, and delete it if it does

mahditaani replied:
Done, see cbd79d33abcdca16a32d19fd8ae3294921acd240

tdealtry commented on src/WCSimWLSProperties.cc:
```diff
@@ -0,0 +1,379 @@
+  //       0.9999, 0.9999, 0.99, 0.97, 0.89, // 350 --- 300
+  //       0.71, 0.54, 0, 0, // 300 --- 250
+  //       0, 0, 0, 0, 0}; // 250 --- 200
+  gAbs = new TGraph();
```
It's more efficient to allocate the TGraph space required, rather than it having to do it multiple times as it's filled
```suggestion
  gEm = new TGraph(nEntries_abslength_eljen);
```
Also, please check if the graph exists, and delete it if it does

mahditaani replied:
Done, see cbd79d33abcdca16a32d19fd8ae3294921acd240

tdealtry commented on src/WCSimWLSProperties.cc:
```diff
@@ -0,0 +1,379 @@
+  //       0.71, 0.54, 0, 0, // 300 --- 250
+  //       0, 0, 0, 0, 0}; // 250 --- 200
+  gAbs = new TGraph();
+  const int step = 10;
```
```suggestion
```
Don't think this is used

mahditaani replied:
You're right. Good spot.

tdealtry commented on include/WCSimWLSProperties.hh:
```diff
@@ -0,0 +1,58 @@
+  G4double* GetPhotonEnergy_EM();
+  G4double* GetEm();
+
+  void SetgAbs();
+  void SetgEm();
+
+  void SethEm();
```
```suggestion
 private:
  void SetgAbs();
  void SetgEm();

  void SethEm();
```

tdealtry commented on src/WCSimRunAction.cc:
```diff
@@ -103,6 +103,12 @@ void WCSimRunAction::BeginOfRunAction(const G4Run* /*aRun*/)
   wcsimrandomparameters->SaveOptionsToOutput(wcsimrootoptions);
+
+  //set Tree for gathering photon info inside stepping action
+  photonTree = new TTree("photons","Photon Tree");
```
This looks like a nice feature, but we need to be careful about bloating file sizes. Can you provide some file size information?
I'd suggest adding an boolean option to save this in `WCSimRunActionMessenger`

tdealtry replied:
Ok so I just run 10 events with OD.mac, and this is huge!
- Photon tree: 38 MB
- wcsimT tree: 1 MB

So we defintiely need to switch this off by default. I'd suggest something like 
- add a `bool` option to `WCSimRunActionMessenger` called `SaveWLSPhotonInfo`
- only create the `TTree` here if that's `true`
- only `Fill()` the tree in `WCSimSteppingAction.cc` if the tree is not `null`

It'd also be good to improve the description e.g.
```suggestion
  photonTree = new TTree("photons","Photons in WLS Tree");
```

mahditaani replied:
@tdealtry thanks for your suggestions but actually this feature is no longer needed. It was initially used to create the wavelength shifting process but we have now moved over to a completely geant4 implementation. As such I have removed this and other unused code in 78f458ebf883fc3a6652906b4315fa3c4cfcbaf7 .

tdealtry commented on src/WCSimConstructMaterials.cc:
```diff
@@ -180,6 +181,28 @@ void WCSimDetectorConstruction::ConstructMaterials()
+  WLS_PVT->AddElement(elH, 10);
+
+
+  //**cladding properties
```
Same comment about renaming this as being explicitly OD WLS cladding, rather than any other form of cladding

tdealtry commented on src/WCSimConstructMaterials.cc:
```diff
@@ -838,11 +865,105 @@ void WCSimDetectorConstruction::ConstructMaterials()
+
+  // MATERIAL properties
+  EljenEJ286 *WLSProps = new EljenEJ286();
+  // Define normal reflectivity from Fresnel equations
+  const G4int nEntries_transmittance = 33;
+  G4double WLS_transmittance_vs_energy[nEntries_transmittance] =
+      { 1-0.020836, 1-0.0207796, 1-0.0207198, 1-0.0206584, 1-0.0205953,
+        1-0.0205288, 1-0.0204589, 1-0.0203874, 1-0.0203125, 1-0.0202344,
+        1-0.0201512, 1-0.0200647, 1-0.019975, 1-0.0198787, 1-0.0197774,
+        1-0.0196696, 1-0.0195553, 1-0.0194329, 1-0.0193041, 1-0.0191639,
+        1-0.019016, 1-0.0188553, 1-0.0186803, 1-0.0184931, 1-0.0182871,
+        1-0.0180644, 1-0.0178189, 1-0.0175494, 1-0.0172519, 1-0.0169209,
+        1-0.0165512, 1-0.0161367, 1-0.0156689};
+  WlsPlasticMPT->AddProperty("TRANSMITTANCE", WLSProps->GetPhotonEnergy(), WLS_transmittance_vs_energy, nEntries_transmittance);
```
Does it make sense for this to go into `WCSimWLSProperties`, like the other WLS properties?

mahditaani replied:
Yes that is a good suggestion. I have moved it.

tdealtry commented on include/WCSimDetectorConstruction.hh:
```diff
@@ -309,6 +371,7 @@ private:
   // ############################# //
 
   bool isODConstructed;
+  bool isCombinedPMTCollectionDefined;
```
This isn't initialised in the constructor to `false`. It should be

tdealtry commented on include/WCSimDetectorConstruction.hh:
```diff
@@ -337,6 +402,10 @@ private:
 
+  // WLS material name
+  bool isWLSFilled;
+  bool BuildCladding;
```
Please label as OD WLS cladding

tdealtry commented on include/WCSimDetectorMessenger.hh:
```diff
@@ -45,6 +46,25 @@ class WCSimDetectorMessenger: public G4UImessenger
+  // isWLS constructed
+  G4UIcmdWithoutParameter* isWLSFilled;
+  // build cladding around WLS
+  G4UIcmdWithoutParameter* BuildCladding;
```
Same comment - please label as OD WLS cladding

tdealtry commented on src/WCSimDetectorMessenger.cc:
```diff
@@ -107,6 +107,119 @@ WCSimDetectorMessenger::WCSimDetectorMessenger(WCSimDetectorConstruction* WCSimD
+
+  // Nb of OD PMT per cell HORIZONTALLY
+  PMTODperCellHorizontal = new G4UIcmdWithAnInteger("/WCSim/HyperKOD/PMTODperCellHorizontal", this);
+  PMTODperCellHorizontal->SetGuidance("Set number of OD PMT per cell HORIZONTALLY (unit: cm mm).");
```
```suggestion
  PMTODperCellHorizontal->SetGuidance("Set number of OD PMT per cell HORIZONTALLY.");
```

mahditaani replied:
Yup, silly mistake. Good catch. Fixed: 3722b24dd5ba65e301dc282f56908ab9b549e09b

tdealtry commented on src/WCSimDetectorMessenger.cc:
```diff
@@ -107,6 +107,119 @@ WCSimDetectorMessenger::WCSimDetectorMessenger(WCSimDetectorConstruction* WCSimD
+
+  // Nb of OD PMT per cell HORIZONTALLY
+  PMTODperCellVertical = new G4UIcmdWithAnInteger("/WCSim/HyperKOD/PMTODperCellVertical", this);
+  PMTODperCellVertical->SetGuidance("Set number of OD PMT per cell VERTICALLY (unit: cm mm).");
```
```suggestion
  PMTODperCellVertical->SetGuidance("Set number of OD PMT per cell VERTICALLY.");
```

mahditaani replied:
Yup, silly mistake. Good catch. Fixed: 3722b24dd5ba65e301dc282f56908ab9b549e09b

tdealtry commented on src/WCSimDetectorMessenger.cc:
```diff
@@ -107,6 +107,119 @@ WCSimDetectorMessenger::WCSimDetectorMessenger(WCSimDetectorConstruction* WCSimD
+
+  // Nb of OD PMT per cell HORIZONTALLY
+  PMTODPercentCoverage = new G4UIcmdWithADouble("/WCSim/HyperKOD/PMTODPercentCoverage", this);
+  PMTODPercentCoverage->SetGuidance("Set global OD photocoverage percentage (unit: cm mm).");
```
```suggestion
  PMTODPercentCoverage->SetGuidance("Set global OD photocoverage percentage.");
```

mahditaani replied:
Yup, silly mistake. Good catch. Fixed: 3722b24dd5ba65e301dc282f56908ab9b549e09b

tdealtry commented on src/WCSimEventAction.cc:
```diff
@@ -867,8 +865,7 @@ void WCSimEventAction::FillRootEvent(G4int event_id,
   int k;
-  //Modify to add decay products
-  for (k=0;k<jhfNtuple.npar;k++) // should be just 2
+  for (k=0;k<2;k++) // should be just 2
```
Please revert this

tdealtry commented on src/WCSimEventAction.cc:
```diff
@@ -982,8 +977,6 @@ void WCSimEventAction::FillRootEvent(G4int event_id,
       } else if (pionList.count(trj->GetParentID()) ) {
 	parentType = 211;
-      } else if (primaryList.count(trj->GetParentID()) ) {
-	parentType = 1;
```
Please revert this

tdealtry commented on src/WCSimEventAction.cc:
```diff
@@ -908,7 +905,6 @@ void WCSimEventAction::FillRootEvent(G4int event_id,
   // same, april 7th 2005
   std::set<int> pionList;
   std::set<int> antipionList;
-  std::set<int> primaryList;
```
Please revert this

tdealtry commented on src/WCSimEventAction.cc:
```diff
@@ -939,7 +935,6 @@ void WCSimEventAction::FillRootEvent(G4int event_id,
     if ( trj->GetPDGEncoding() == 211 ) pionList.insert(trj->GetTrackID());
     if ( trj->GetPDGEncoding() == -211 ) antipionList.insert(trj->GetTrackID());
        
-    if( trj->GetParentID() == 0 ) primaryList.insert(trj->GetTrackID());
```
Please revert this

tdealtry commented on src/WCSimPMTObject.cc:
```diff
@@ -1965,3 +2011,46 @@ G4double PMT3inch::GetDarkRateConversionFactor(){
+}
+
+WCSimBasicPMTObject::~WCSimBasicPMTObject(){
+}
```
Please cleanup the TGraph

tdealtry commented on src/WCSimPMTObject.cc:
```diff
@@ -1965,3 +2011,46 @@ G4double PMT3inch::GetDarkRateConversionFactor(){
+}
+
+void WCSimBasicPMTObject::DefineQEHist(std::map<G4double,G4double> mapQE){
+  gQE = new TGraph();
```
```suggestion
  gQE = new TGraph(mapQE.size());
```
should be more efficient.
Please also check if the TGraph exists

tdealtry commented on include/WCSimPMTObject.hh:
```diff
@@ -6,7 +6,8 @@
 #include <map>
 #include <vector>
-
+#include <TH1F.h>
```
```suggestion
```
not used

mahditaani replied:
Removed. See efaef16e74a05b14a5ced84158e3ae083a71ee8a and 2d485754aab0323152f3f7896280c07f24dd4eef

tdealtry commented on src/WCSimPMTObject.cc:
```diff
@@ -1965,3 +2011,46 @@ G4double PMT3inch::GetDarkRateConversionFactor(){
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////
+// COMBINED PMTS CLASS
+// Goal is to instantiate purely virtual WCSimPMTObject Classes to contained an extended
+//  QE information of several PMTs collection
+// Useful for applying QE in stacking action when several PMTs collection are defined
```
- In the ID-only case, this combined class isn't used, and the QE values are taken from the PMT class directly
- In the ID+OD case, this combined class is used, and the QE values are taken from a combination of ID & OD PMTs QE values inside `WCSimDetectorConstruction::CreateCombinedPMTQE()`

My reading of this is that there's a potential for differences in the ID response between running and ID-only & and ID+OD geometry. This is of course something we'd like to avoid. Am I missing something?

mahditaani replied:
@federiconova do you have any comments on this?

federiconova replied:
Hi, sorry, I am happy to comment but what are the incriminated lines? I only see four commented lines.

mahditaani replied:
I see that in `WCSimDetectorConfigs` we make use of the combined PMT QE for the OD configuration. Does this affect the QE of the ID PMTs at all?

federiconova replied:
Mahdi I think you had implemented these lines but they are actually never used. In DetectorConfigs, isCombinedPMTCollectionDefined=true. So then in StackingAction the code uses GetPMTQE instead of GetStackingPMTQE. So the WCSimBasicPMT object is then never used (in WCSimPMTQE). I feel it should be safe to just remove the BasicPMT class (to which these comments refer) and all references to it.

tdealtry replied:
Ok thanks for this. Removing the `WCSimBasicPMTObject` class sounds great

tdealtry commented on src/WCSimPMTQE.cc:
```diff
@@ -38,7 +38,7 @@ G4double WCSimDetectorConstruction::GetPMTQE(G4String CollectionName, G4double P
   // return 0 for wavelenght outside the range
   if (flag==1){
-    if (PhotonWavelength <= low_wl || PhotonWavelength >= high_wl || PhotonWavelength <=280 || PhotonWavelength >=660){
+    if (PhotonWavelength <= low_wl || PhotonWavelength >= high_wl ){
```
This is a potential cause of the difference seen from the Travis CI tests. Please revert for now so we can test

mahditaani replied:
I guess this is an optimisation for ID PMTs but the OD will use WLS plates that can detect photons with less than 280nm wavelength. I will revert but this will need to be included for the OD to function correctly.

federiconova replied:
That's right. But anyway, why is the test failing? If extending the possible wavelengths changes the results for the ID, doesn't that mean that they *should* be extended, because some photons are being missed?

tdealtry commented on include/WCSimSteppingAction.hh:
```diff
@@ -1,19 +1,35 @@
 #ifndef WCSimSteppingAction_h
 #define WCSimSteppingAction_h 1
 
+#include <TTree.h>
```
```suggestion
```

mahditaani replied:
Removed. See efaef16e74a05b14a5ced84158e3ae083a71ee8a and 2d485754aab0323152f3f7896280c07f24dd4eef

tdealtry commented on src/WCSimSteppingAction.cc:
```diff
@@ -13,6 +17,66 @@
+// uses only the position of the photon's absorption with respect to the PMT (distance from PMT) and
+// lab measurements of the efficiency as a function of distance, taken at Edinburgh University.
+bool EmittedPhotonDetected(G4ThreeVector absorbedPhotonPosition){
+  TRandom3 *r = new TRandom3(time(0)); // This can be changed to whatever random number generator WCSim is using.
```
Creating a new `TRandom3` everytime this is called will give a big memory leak.
If you want to keep a separate random number generator just for this process, please create it in the constructor (and delete it in the destructor), and come up with an appropriate way to set the seed (e.g. in `WCSimRandomParameters::SetSeed()`)
Otherwise `G4UniformRand()` should work

tdealtry commented on src/WCSimSteppingAction.cc:
```diff
@@ -13,6 +17,66 @@
+const double planck = 4.1357e-15; // ev.s
+const double lightspeed = 299792458e9; // nm/s
+
+// This function will determine the probability of an emitted photon will being detected by the PMT.
```
Please make clear that these are WLS-related (or even WLS+PMT-system-related) functions

mahditaani replied:
I think I will remove these functions in another pull request as they aren't used any more.

tdealtry commented on src/WCSimSteppingAction.cc:
```diff
@@ -27,17 +91,14 @@ void WCSimSteppingAction::UserSteppingAction(const G4Step* aStep)
+  WLS=true;
+  if(WLS){
+    CollectPhotonsInfoInWLSPlates(aStep);
+    // WLSPhysicsProcess(aStep);
```
Why is this commented? Is the method even required anymore?

mahditaani replied:
Yeah this process is no longer required but I will remove them in the next pull request.

tdealtry commented on src/WCSimSteppingAction.cc:
```diff
@@ -27,17 +91,14 @@ void WCSimSteppingAction::UserSteppingAction(const G4Step* aStep)
-  
+  // WLS
+  bool WLS=false;
+  WLS=true;
```
A better way would be to set this to `true` only if there is WLS in the geometry

tdealtry commented on src/WCSimSteppingAction.cc:
```diff
@@ -160,3 +221,228 @@ double WCSimSteppingAction::FieldLines(G4double /*x*/,G4double /*y*/,G4int /*coo
+  // We don't know the spatial distribution of the emitted photon so we will just keep this distribution uniform
+  // over the surface of a sphere.
+
+  TRandom3 *r = new TRandom3(time(0)); // This can be changed to whatever random number generator WCSim is using.
```
Another memory leak. Please fix

tdealtry commented on src/WCSimStackingAction.cc:
```diff
@@ -26,32 +26,40 @@ G4ClassificationOfNewTrack WCSimStackingAction::ClassifyNewTrack
+      // MF : translated from skdetsim : better to increase the number of photons
+      // than to throw in a global factor  at Digitization time !
+      //G4double ratio = 1./(1.0-0.25);
+      G4double ratio = 0.5;
```
The ratio has been changed from 4/3 to 1/2. Why?
This could be the cause of the Travis CI differences. Please revert for now so we can check

mahditaani replied:
Looking through the code, it seems that this just arbitrarily increases the QE. I am not sure I understand why this is even here. @federiconova if I remember correctly you had a reason for changing this. In any case, I will revert for now.

federiconova replied:
ok, thanks for this work!

tdealtry commented on src/WCSimStackingAction.cc:
```diff
@@ -26,32 +26,40 @@ G4ClassificationOfNewTrack WCSimStackingAction::ClassifyNewTrack
 	  if (DetConstruct->GetPMT_QE_Method() == 1){
-	    wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,1,240,660,ratio);
+	    if(!DetConstruct->GetIsCombinedPMTCollectionDefined())
+	      wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,1,200,700,ratio);
```
```suggestion
	      wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,1,240,660,ratio);
```

tdealtry commented on src/WCSimStackingAction.cc:
```diff
@@ -26,32 +26,40 @@ G4ClassificationOfNewTrack WCSimStackingAction::ClassifyNewTrack
       if(aTrack->GetCreatorProcess()==NULL) {
-	wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,1,240,660,ratio);
+	if(!DetConstruct->GetIsCombinedPMTCollectionDefined())
+	  wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,1,200,700,ratio);
```
```suggestion
	  wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,1,240,660,ratio);
```
Another change that could be in the Travis CI. And seemingly, this one doesn't affect the OD. So please revert

tdealtry commented on src/WCSimStackingAction.cc:
```diff
@@ -26,32 +26,40 @@ G4ClassificationOfNewTrack WCSimStackingAction::ClassifyNewTrack
+	      wavelengthQE  = DetConstruct->GetStackingPMTQE(photonWavelength,1,200,700,ratio);
 	  }else if (DetConstruct->GetPMT_QE_Method() == 2){
+	    if(!DetConstruct->GetIsCombinedPMTCollectionDefined())
+	      wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,0,200,700,ratio);
```
```suggestion
	      wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,0,240,660,ratio);
```

tdealtry commented on src/WCSimStackingAction.cc:
```diff
@@ -26,32 +26,40 @@ G4ClassificationOfNewTrack WCSimStackingAction::ClassifyNewTrack
+	      wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,0,200,700,ratio);
+	    else
+	      wavelengthQE  = DetConstruct->GetStackingPMTQE(photonWavelength,0,200,700,ratio);
 	    wavelengthQE  = DetConstruct->GetPMTQE(WCIDCollectionName,photonWavelength,0,240,660,ratio);
```
Did you mean to delete this line?

mahditaani replied:
Yes it would seem so!

tdealtry commented on src/WCSimDetectorConfigs.cc:
```diff
@@ -243,15 +243,16 @@ void WCSimDetectorConstruction::SetHyperKWithODGeometry()
   WCPMTRadius         = PMT->GetRadius();
-  WCIDDiameter          = 70.8*m; // = 74m - 2*(60cm ID wall + 1m OD)
-  WCIDHeight            = 54.8*m; // = 60m - 2*(60cm ID wall + 2m OD)
+  WCIDDiameter          = 64.8*m; // = 68m - 2*(60cm ID wall + 1m OD)(69m with 0.5m concrete each side)
+  WCIDHeight            = 65.8*m; // = 71m - 2*(60cm ID wall + 2m OD)(72m with 0.5m concrete on each side)
```
Are these the official "this is how big the HK tank will be" numbers? Do you have a reference for them?
(If these numbers are correct, I'll sort out the ID-only geometry. It should of course be the same as the ID part of this geometry)

mahditaani replied:
Yeah these numbers were shown several collaboration meetings ago and are the numbers now stated in the technical report.  I will note that changing to the new numbers caused some stuck tracks but I can't figure out where they are coming from (I couldn't find any overlaps).

tdealtry commented on src/WCSimDetectorConstruction.cc:
```diff
@@ -297,3 +298,131 @@ void WCSimDetectorConstruction::SaveOptionsToOutput(WCSimRootOptions * wcopt)
+  std::map<G4double,G4double> QE;
+
+  // Recursive algorithm to set new QE for combined PMT collection
+  // F. Nova: define QE for a given wavelength as sqrt(QE1^2 + QE2^2 + ...), where each QEi is the value of QE for PMT type i
+  // replace it with maximum
```
Please clarify the comment - looks like `max()` rather than `sqrt()` is used now

mahditaani replied:
Hi @federiconova do you have any thoughts on this?

federiconova replied:
Hi, is this function used at all? I suspect it isn't.
I think initially it was defined as sqrt and then I replaced it with max, but eventually we are not even using it?

mahditaani replied:
@federiconova on line 376 the std::max() function is used. Am I correct in saying that the new QE of a combined collection is the highest QE for that wavelength?

federiconova replied:
Yes.

tdealtry commented on src/WCSimDetectorConstruction.cc:
```diff
@@ -297,3 +298,131 @@ void WCSimDetectorConstruction::SaveOptionsToOutput(WCSimRootOptions * wcopt)
+          G4cout << " ### " << it->first << "nm : " << it->second << G4endl;
+          QE[it->first]=it->second;
+        } else {
+	  //	  max_QE = sqrt(it->second*it->second + foundWL->second*foundWL->second);
```
```suggestion
```

mahditaani replied:
Removed. See efaef16e74a05b14a5ced84158e3ae083a71ee8a and 2d485754aab0323152f3f7896280c07f24dd4eef

tdealtry commented on sample-root-scripts/llib.C:
```diff
@@ -0,0 +1,16 @@
+void llib(){
```
This file isn't required anymore. Please use `$WCSIMDIR/rootwc` instead to automatically load WCSim libraries & includes

tdealtry commented on sample-root-scripts/ODAnalysis.C:
```diff
@@ -0,0 +1,329 @@
+
+
+/*
+* To run this script just type:  root -l -x llib.C 'ODAnalysis.C("wcsim.root", "ODHits.root", false)' 
```
```suggestion
* To run this script just type:  rootwc -l -x 'ODAnalysis.C("wcsim.root", "ODHits.root", false)' 
```

tdealtry commented on sample-root-scripts/ODAnalysis.C:
```diff
@@ -0,0 +1,329 @@
+* To run this script just type:  root -l -x llib.C 'ODAnalysis.C("wcsim.root", "ODHits.root", false)' 
+* or replace wcsim.root and ODHits.root for your input and output filenames respectively
+* The final true can be changed to true to turn on verbosity.
+* The default: root -l -x llib.C ODAnalysis.C will run with the variables mentioned above
```
```suggestion
* The default: root -l -x ODAnalysis.C will run with the variables mentioned above
```

tdealtry commented on sample-root-scripts/ODAnalysis.C:
```diff
@@ -0,0 +1,329 @@
+	}
+
+
+}// End of loadlibs function
```
```suggestion
```

tdealtry commented on sample-root-scripts/ODAnalysis.C:
```diff
@@ -0,0 +1,329 @@
+
+void ODAnalysis( const char *inFileName = "wcsim.root", const char *outFileName = "ODHits.root", bool verbosity = 0){ 
+	
+//	loadlibs(); // Load the required libraries
```
```suggestion
```

tdealtry commented on sample-root-scripts/ODAnalysis.C:
```diff
@@ -0,0 +1,329 @@
+	TTree *wcsimTree = (TTree*) inFile->Get("wcsimT");
+	
+	// Get the number of events in the tree
+	long int nEvent = wcsimTree->GetEntries();
```
```suggestion
	const long int nEvent = wcsimTree->GetEntries();
```

tdealtry commented on sample-root-scripts/ODAnalysis.C:
```diff
@@ -0,0 +1,329 @@
+
+
+
+	for (int ev = 0; ev < nEvent; ev++){ // Loop over events
```
```suggestion
	for (long int ev = 0; ev < nEvent; ev++){ // Loop over events
```

tdealtry commented on sample-root-scripts/trigger_OD.C:
```diff
@@ -0,0 +1,201 @@
+  /* A simple script to plot aspects of phototube hits 
+   * 
+   * I like to run this macro as 
+   * $ root -l -x 'trigger_OD.C("OD.root")'
```
```suggestion
   * $ rootwc -l -x 'trigger_OD.C("OD.root")'
```

mahditaani replied:
I am unable to use this rootwc command after building with cmake. How is it supposed to be used?

mahditaani replied:
Never mind, it is because it isn't automatically copied when building with cmake.

tdealtry commented on sample-root-scripts/trigger_OD.C:
```diff
@@ -0,0 +1,201 @@
+    gSystem->Load("${WCSIMDIR}/src/WCSimRootDict_rdict.pcm");
+  }else{
+    std::cout << "Can't load WCSim ROOT dictionaries" << std::endl;
+  }
```
```suggestion
```

tdealtry commented on sample-root-scripts/ODAnalysis.C:
```diff
@@ -0,0 +1,329 @@
+// C++ Includes
```
From a quick scan, it looks like this file and `read_OD.C` are very similar, with the major differences being
- ODAnalysis saves plotted 2D histogram canvases, in addition to the TTree
- read_OD saves information from ID & OD, rather than just the OD
Is it possible to merge them into one macro? There could be `bool plot` and `bool saveID` arguments

mahditaani replied:
You are right. I have removed the redundant macro.

tdealtry reviewed (changes requested):
Here's my other set

There's some ideas of what's going on with the Travis plots. Hopefully we can see agreement after some reversions, then talk about which changes are relevant for ID+OD, and which should be OD-only

mahditaani committed:
Copied data files and OD macros to build directory.

mahditaani committed:
Removed unnecessary OD macro

mahditaani committed:
Update src/WCSimTuningParameters.cc

Set the default OD Tyvek reflectivity to 90% in line with the rest of the code.

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

mahditaani committed:
Merge branch 'prod/ODWithWLS' of github.com:mahditaani/WCSim into prod/ODWithWLS

mahditaani committed:
Added comments to OD mac files.

mahditaani committed:
Added correct default value for cosmics data

mahditaani committed:
Added a comment to show the option to change the cosmics data file in WCSim.mac

mahditaani committed:
Changed name of output cosmic flux histogram file.

mahditaani committed:
Renamed WLS Cladding reflectivity parameter for clarity

mahditaani committed:
Removed old seed generator command. Now set using /WCSim/random/seed

mahditaani committed:
Added comment to WCSim.mac for clarity on the OD options.

mahditaani committed:
Removed unnecessary code/comments

mahditaani committed:
Removed inaccurate guidance text for OD mac options.

mahditaani committed:
Removed unnecessary lines of code.

mahditaani committed:
Actually removed unnecessary lines of code.

tdealtry commented:
Thanks for this @mahditaani & @federiconova I'll have a proper look tomorrow (github actions have sponged all my time today... but look to have something working now. Will close/open this PR to trigger a build)

mahditaani committed:
Update src/WCSimTrackingAction.cc

Revert change.

Co-authored-by: Tom Dealtry <tom.deal135@gmail.com>

mahditaani committed:
Revert changes in jhfNtuple to before the PR ( commit e5d51e71612d6a9fdb69a425518a519ef8450aad )

mahditaani committed:
Added lines to prevent seg faults from unassigned targetpdg variable.

mahditaani committed:
Changed wavelength range of photons back to 240 - 660.

mahditaani committed:
Renamed BuildCladding to clearer BuildODWLSCladding

mahditaani committed:
Reverted wavelength check to pass CI tests.

tdealtry commented:
Doing another close/open to check if the permissions for uploading to the validation website are fixed

tdealtry commented:
Ok, so no idea what's happened there. It seems to have run the validation on WCSim/develop rather than mahditaani/prod/ODWithWLS. I think I've got a fix, so will do another close/open (but of course it could be that a close/open by me isn't the same as a push by you...)
On the plus side, the action is now much faster, now we're not simulating dark noise for a **long** time in the validation

tdealtry commented:
I'm happy the validation is correctly working on your branch now, but the output is still different to what's expected unfortunately... will delve next week

mahditaani committed:
Reverted changes in WCSimEventAction.cc. Reintroduced the priimaryList.

mahditaani committed:
Reverted wavelengths to old 240-660 range and changed QE ratio back to 4/3.

mahditaani committed:
Fixed memory leak in WLSProperties

mahditaani committed:
Copy rootwc to build directory

mahditaani committed:
Included the new rootwc command in the analysis scripts.

mahditaani committed:
Set number of events to const in od script.

mahditaani committed:
Fixed memory leak in Stepping Action (WLS)

mahditaani committed:
Fixed potential memory leak in PMTObject

mahditaani committed:
Initialised variable in detector construction.

mahditaani committed:
Renamed WLS OD cladding properties for clarity.

mahditaani committed:
Made Eljen plate WLS functions private.

mahditaani committed:
Moved WLS Property to appropriate class.

mahditaani committed:
Clarified combined PMT collection comment

mahditaani committed:
Removed unused variables from WLSProperties

mahditaani committed:
Modified OD macro dark rate options for clarity.

mahditaani committed:
Removed redundant script.

mahditaani committed:
Removed computing scripts from this branch. Find them in prod/ODWithWLSWithComputing.

tdealtry commented on WCSim.cc:
```diff
@@ -1,3 +1,4 @@
+#include <TRandom3.h>
```
```suggestion
```
Not required

tdealtry commented on src/WCSimPrimaryGeneratorAction.cc:
```diff
@@ -111,6 +174,7 @@ WCSimPrimaryGeneratorAction::~WCSimPrimaryGeneratorAction()
+  inputCosmicsFile.close();
   delete particleGun;
   delete MyGPS;   //T. Akiri: Delete the GPS variable
   delete messenger;
```
```suggestion
  delete messenger;
  delete hFluxCosmics;
  delete hEmeanCosmics;
```
Noitced a couple of histograms that should be cleared up in the destructor

mahditaani replied:
Good spot! done, see 4fc0e54e2a93c7acb5fb2ac65c143299335982fc

tdealtry commented on src/WCSimEventAction.cc:
```diff
@@ -1002,7 +999,7 @@ void WCSimEventAction::FillRootEvent(G4int event_id,
 	stop[l]=Stop[l]/cm; // stopping point 
 	start[l]=Start[l]/cm; // starting point 
-	//G4cout<<"part 2 start["<<l<<"]: "<< start[l] <<G4endl;
+	G4cout<<"part 2 start["<<l<<"]: "<< start[l] <<G4endl;
```
Please comment this out again (or any other way where it's off by default). It's very verbose!
```suggestion
	//G4cout<<"part 2 start["<<l<<"]: "<< start[l] <<G4endl;
```

tdealtry commented on src/WCSimEventAction.cc:
```diff
@@ -880,7 +877,7 @@ void WCSimEventAction::FillRootEvent(G4int event_id,
       stop[l]=jhfNtuple.stop[k][l];
       start[l]=jhfNtuple.start[k][l];
-      //G4cout<< "start[" << k << "][" << l <<"]: "<< jhfNtuple.start[k][l] <<G4endl;
+	G4cout<< "start[" << k << "][" << l <<"]: "<< jhfNtuple.start[k][l] <<G4endl;
```
Please comment this out again (or any other way where it's off by default). It's very verbose!
```suggestion
	//G4cout<< "start[" << k << "][" << l <<"]: "<< jhfNtuple.start[k][l] <<G4endl;
```

tdealtry reviewed (changes requested):
Thanks for all this @mahditaani 
I've had a skim through it all again, and just a few comments left (most trivial, couple are a bit more work)
Almost time to get this PR merged in!

mahditaani committed:
Removed unnecessary header file

mahditaani committed:
Commented out printout to reduce verbosity

mahditaani committed:
Commented out printout to reduce verbosity.

mahditaani committed:
Clean up histograms in WCSimPrimaryGeneratorAction destructor.

mahditaani committed:
Removed unused WLS processes. Moved to complete Geant4 implementation.

tdealtry commented:
Thanks for all this work @mahditaani and the rest of the OD group!
